File: src/main/java/com/hotakademy/dockwitcher/domain/repositories/AbstractMongoRepository.java
package com.hotakademy.dockwitcher.domain.repositories;


import com.hotakademy.dockwitcher.domain.entities.Entity;
import org.bson.types.ObjectId;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.data.mongodb.core.query.Criteria;
import org.springframework.data.mongodb.core.query.Query;

import java.util.Optional;

public class AbstractMongoRepository<TEntity extends Entity> implements IMongoRepository<TEntity> {

    protected MongoTemplate mongoTemplate;
    protected Class<TEntity> entityClass;

    public AbstractMongoRepository(MongoTemplate mongoTemplate, Class<TEntity> entityClass) {
        this.mongoTemplate = mongoTemplate;
        this.entityClass = entityClass;
    }

    @Override
    public void save(TEntity entity){
        mongoTemplate.save(entity);
    }

    @Override
    public Optional<TEntity> findById(ObjectId objectId){
        return Optional.ofNullable(mongoTemplate.findById(objectId, entityClass));
    }

    @Override
    public boolean existsById(ObjectId id){
        Query query = new Query();
        query.addCriteria(Criteria.where("_id").is(id));
        return mongoTemplate.exists(query, entityClass);
    }
}

File: src/main/kotlin/com/hotakakademy/dockwitcher/domain/repositories/AbstractMongoRepository.kt
package com.hotakakademy.dockwitcher.domain.repositories

import com.hotakakademy.dockwitcher.domain.entities.Entity
import org.bson.types.ObjectId
import org.springframework.data.mongodb.core.MongoTemplate
import org.springframework.data.mongodb.core.query.Criteria
import org.springframework.data.mongodb.core.query.Query
import java.util.*

open class AbstractMongoRepository<TEntity : Entity>(protected var mongoTemplate: MongoTemplate, protected var entityClass: Class<TEntity>) : IMongoRepository<TEntity> {
    override fun save(entity: TEntity) {
        mongoTemplate.save(entity)
    }

    override fun findById(objectId: ObjectId?): TEntity? {
        return mongoTemplate.findById(objectId!!, entityClass)
    }

    override fun existsById(id: ObjectId?): Boolean {
        val query = Query()
        query.addCriteria(Criteria.where("_id").`is`(id))
        return mongoTemplate.exists(query, entityClass)
    }
}